### PR TITLE
Update the default detach semantic for forward mode AD

### PR DESCRIPTION
**Pull request by albanD**
This makes detach both forward and backward non-differentiable by default.
You can pass the `only_backward_mode=True` argument to make it forward differentiable but backward non-differentiable.

The important side effect of this change is that, by default, detach is not tracking any view information.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3571,15 +3571,15 @@
 # to false to make such changes explicitly illegal, in order to prevent users from
 # changing metadata of the detached tensor and expecting the original tensor to also
 # be updated.
-- func: detach(Tensor(a) self) -> Tensor(a)
+- func: detach(Tensor(a) self, bool only_backward_mode=False) -> Tensor(a)
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:3574]: make this kwarg only?

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:3574]: Ho we can do that that here? 
Will do thanks for the comment! :)

**Comment by ezyang** [comment]
After seeing this PR, I changed my mind about the extra boolean argument on detach. I'd prefer if we just provided a separate functions for the disjoint cases (detach backward only; detach forward only).

**Comment by ezyang** [comment]
It probably will also help reduce BC blast radius

**Comment by ezyang** [comment]
> Well actually, the current state of .detach() is the one that is BC-breaking. In the sense that now detach is a differentiable view (for forward) and so it starts tracking more things than it used to.
This new default, would make detach pretty much the exact same thing from 1.8 to 1.9.

What I mean is, adding new defaulted arguments to functions *should* be unlikely to break things, but in practice it isn't because we hardcode the arguments into serialized models and that breaks forwards compatibility. Less stuff is likely to go wrong if you put it in another method, and I also think this is better, because the vast majority of users who use `detach()` don't care about this argument and you shouldn't force them to think about it, e.g., when they pull up the docs.

I don't care about forward AD implications, no one is using forward AD, let's change things because it's right.

**Comment by ezyang** [comment]
And also, like, other BC incompatibilities from adding a new defaulted argument include people who are explicitly referring to the full schema string (they need to be updated) and external backends who implement this function (they are broken this way)

**Comment by albanD** [comment]
@ezyang ok I updated the existing detach to have the new behavior.
We can discuss naming for the new functions in a follow up.

**Review by ezyang — APPROVED** [review]

**Comment by min-xu-ai** [comment]
> You can pass the only_backward_mode=True argument to make it forward differentiable but backward non-differentiable.

Is the above still true in the committed version? I didn't fully get the reason for doing this. Is it to add more flexibility to `detach`?
